src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,348 @@
+  params.addParam<Real>("rebar_radius", 1.0, "Radius of the rebar");
+  return params;
+}
+
+RebarBondSlipConstraint::RebarBondSlipConstraint(const InputParameters & parameters)
+  : EqualValueEmbeddedConstraint(parameters),
+    _component(getParam<unsigned int>("component")),
+    _mesh_dimension(_mesh.dimension()),
+    _var_nums(_mesh_dimension, libMesh::invalid_uint),
+    _vars(_mesh_dimension, nullptr),
+    _debug(getParam<bool>("debug")),
+    _max_bondstress(getParam<Real>("max_bondstress")),
+    _frictional_bondstress(getParam<Real>("frictional_bondstress")),
+    _ultimate_slip(getParam<Real>("ultimate_slip")),
+    _bar_radius(getParam<Real>("rebar_radius")),
+    _transitional_slip(getParam<std::vector<Real>>("transitional_slip_values"))
+{
+  if (_mesh_dimension != coupledComponents("displacements"))
+    mooseError("In RebarBondSlipConstraint, number of displacements must equal the mesh dimension");
+
+  for (unsigned int i = 0; i < _mesh_dimension; ++i)
+  {
+    _var_nums[i] = coupled("displacements", i);
+    _vars[i] = getVar("displacements", i);
+  }
+}
+
+void
+RebarBondSlipConstraint::initialSetup()
+{
+  for (auto it = _slave_to_master_map.begin(); it != _slave_to_master_map.end(); ++it)
+    if (_bondslip.find(it->first) == _bondslip.end())
+      _bondslip.insert(std::pair<dof_id_type, bondSlipData>(it->first,
+                                                            bondSlipData())); // initialize
+}
+
+void
+RebarBondSlipConstraint::timestepSetup()
+{
+  for (auto iter = _slave_to_master_map.begin(); iter != _slave_to_master_map.end(); ++iter)
+  {
+    dof_id_type node_id = iter->first;
+    auto it = _bondslip.find(node_id);
+    mooseAssert(it != _bondslip.end(), "Node not found in bond-slip map");
+
+    _bondslip[node_id].slip_min_old = _bondslip[node_id].slip_min;
+    _bondslip[node_id].slip_max_old = _bondslip[node_id].slip_max;
+    _bondslip[node_id].bondstress_min_old = _bondslip[node_id].bondstress_min;
+    _bondslip[node_id].bondstress_max_old = _bondslip[node_id].bondstress_max;
+  }
+}
+
+bool
+RebarBondSlipConstraint::shouldApply()
+{
+  if (_debug)
+  // if (_current_node->id() == 144)
+  {
+    std::cout << "===========================================\n";
+    std::cout << "node id: " << _current_node->id() << std::endl;
+    std::cout << "at coord: " << (Point)*_current_node << std::endl;
+  }
+  auto it = _slave_to_master_map.find(_current_node->id());
+
+  if (it != _slave_to_master_map.end())
+  {
+    const Elem * master_elem = _mesh.elemPtr(it->second);
+    std::vector<Point> points = {*_current_node};
+
+    // reinit variables on the master element at the slave point
+    _fe_problem.setNeighborSubdomainID(master_elem, 0);
+    _fe_problem.reinitNeighborPhys(master_elem, points, 0);
+
+    reinitConstraint();
+
+    return true;
+  }
+  return false;
+}
+
+void
+RebarBondSlipConstraint::computeTangent()
+{
+  _slave_tangent *= 0.0;
```

Comment:
Should be set = 0?

Reply:
I guess that is what this accomplishes.  It just looks strange.  I'd replace it with an =